File: test/Models/Buildable/Installable/StudTest.java
package Models.Buildable.Installable;

import Graphics.GraphicsList;
import Graphics.RectangleInstructions;
import Models.Buildable.Material.Lumber;
import Models.Buildable.Material.MaterialList;
import Models.Buildable.Material.Nail;
import Models.Measurement;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class StudTest {
    Lumber.Dimension dimension = Lumber.Dimension.TWO_BY_FOUR;
    @Test
    public void defaultStudShouldHave8FtLength() {
        MaterialList result = new MaterialList().addMaterial(new Lumber(new Measurement(92, Measurement.Fraction.FIVE_EIGHTH), dimension), 1)
                .addMaterial(Nail.TEN_D, 6);
        MaterialList test = new Stud().materialList();
        String error = String.format("Expected %s, was %s", result.materials(), test.materials());
        assertEquals(result, test, error);
    }

    @Test
    public void studShouldProduceMaterialList() {
        MaterialList result = new MaterialList().addMaterial(Nail.TEN_D, 6)
                .addMaterial(new Lumber(new Measurement(92, Measurement.Fraction.FIVE_EIGHTH), this.dimension), 1);
        assertEquals(result, new Stud().materialList());
        assertEquals(result, new Stud(new Measurement(92), this.dimension).materialList());

        MaterialList secondResult = new MaterialList().addMaterial(Nail.TEN_D, 6)
                .addMaterial(new Lumber(new Measurement(96), this.dimension), 1);
        assertEquals(secondResult, new Stud(new Measurement(92, Measurement.Fraction.ELEVEN_SIXTEENTH), this.dimension).materialList());
    }

    @Test
    public void studShouldBeEqualWhenOfTheSameLength() {
        assertEquals(new Stud(new Measurement(95), this.dimension),
                new Stud(new Measurement(95), this.dimension));
    }

    @Test
    public void studShouldCalculateHashCode() {
        int result = 1742707305;
        assertEquals(result, new Stud().hashCode());
    }

    @Test
    public void studShouldNotBeEqualWhenOfDifferentInstalledLength() {
        assertNotEquals(new Stud(), new Stud(new Measurement(92), this.dimension));
    }

    @Test
    public void studShouldCreateDrawingInstructions() {
        Measurement zero = new Measurement(0);
        Measurement height = new Measurement(10);
        GraphicsList results = new GraphicsList().addGraphic(new RectangleInstructions(zero, zero, dimension.width, height));
        assertEquals(results.drawingInstructions(), new Stud(height, dimension).graphicsList().drawingInstructions());
    }

    @Test
    public void shouldReturnTotalHeight() {
        assertEquals(new Measurement(95), new Stud(new Measurement(95), Lumber.Dimension.TWO_BY_FOUR).totalHeight());
    }
}

File: src/test/kotlin/models/buildable/installable/StudTest.kt
package models.buildable.installable

import graphics.*
import models.buildable.material.*
import models.Measurement
import models.Measurement.Fraction
import kotlin.test.*

internal class StudTest {
    private var dimension = Lumber.Dimension.TWO_BY_FOUR

    @Test
    fun defaultStudShouldHave8FtLength() {
        val result = MaterialList()
            .addMaterial(Lumber(Measurement(92, Fraction.FIVE_EIGHTH), dimension), 1)
            .addMaterial(Nail.TEN_D, 6)
        val test = Stud().materialList()
        assertEquals(result, test)
    }

    @Test
    fun whenLengthIsGreaterThanLumberMaxLengthShouldThrow() {
        val thrown = assertFailsWith<Stud.InvalidLengthException> {
            Stud(Measurement(240, Fraction.ONE_SIXTEENTH))
        }
        assertEquals("cannot be longer than ${Lumber.FactoryLength.maxLength}", thrown.message)
        assertEquals(Lumber.FactoryLength.maxLength, thrown.maxLength)
    }

    @Test
    fun studShouldProduceMaterialList() {
        val result = MaterialList()
            .addMaterial(Nail.TEN_D, 6)
            .addMaterial(Lumber(Measurement(92, Fraction.FIVE_EIGHTH), dimension), 1)
        assertEquals(result, Stud().materialList())
        assertEquals(result, Stud(Measurement(92), dimension).materialList())
        val secondResult = MaterialList()
            .addMaterial(Nail.TEN_D, 6)
            .addMaterial(Lumber(Measurement(96), dimension), 1)
        assertEquals(
            secondResult,
            Stud(Measurement(92, Fraction.ELEVEN_SIXTEENTH), dimension).materialList()
        )
    }

    @Test
    fun studShouldBeEqualWhenOfTheSameLength() {
        assertEquals(Stud(Measurement(95), dimension), Stud(Measurement(95), dimension))
    }

    @Test
    fun studShouldCalculateHashCode() {
        val result = -1875035575
        assertEquals(result, Stud().hashCode())
    }

    @Test
    fun studShouldNotBeEqualWhenOfDifferentInstalledLength() {
        assertNotEquals(Stud(), Stud(Measurement(92), dimension))
    }

    @Test
    fun studShouldCreateDrawingInstructions() {
        val zero = Measurement(0)
        val height = Measurement(10)
        val results = GraphicsList().addGraphic(RectangleInstructions(zero, zero, dimension.width, height))
        assertEquals(
            results.drawingInstructions(),
            Stud(height, dimension).graphicsList().drawingInstructions()
        )
    }

    @Test
    fun shouldReturnTotalHeight() {
        assertEquals(Measurement(95), Stud(Measurement(95), Lumber.Dimension.TWO_BY_FOUR).totalHeight())
    }
}